digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataExporter/AssetReportCsvExporter.scala

```diff
@@ -0,0 +1,264 @@
+
+      if (validUsers.nonEmpty) {
+        val elem = validUsers.head
+        ( elem.modifiedBy , Some(elem.modifiedDate) )
+      }
+      else
+        ("", None)
+    }
+
+    def executeQuery(municipality: Int): List[AssetReport] = {
+      if (extractPointsAsset) {
+        assetReporterDAO.pointAssetQuery(Seq(municipality), assetTypesList)
+      }
+      else {
+        val linkIds = roadLinkServiceImpl.getVVHRoadLinksF(municipality)
+                                         .map(_.linkId)
+
+        assetReporterDAO.linearAssetQuery(linkIds, assetTypesList)
+      }
+    }
+
+    def extractData(municipality: Int): List[AssetReport] = {
+      if (withTransaction) {
+        withDynTransaction {
+          executeQuery(municipality)
+        }
+      }
+      else {
+        executeQuery(municipality)
+      }
+    }
+
+    def getUserAndMunicipalitiesInfo(): (Seq[User], List[MunicipalityInfo]) = {
+      if(withTransaction) {
+        withDynSession {
+          val userResults = userProvider.getUsers()
+          val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+          (userResults, municResults)
+        }
+      }
+      else{
+        val userResults = userProvider.getUsers()
+        val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+        (userResults, municResults)
+      }
+    }
+
+
+    val (users, municipalitiesInfo) = getUserAndMunicipalitiesInfo()
+
+    val operatorUsers = users.filter(_.isOperator())
+    val operatorUsersUsername = operatorUsers.map(_.username)
+
+    municipalitiesList.flatMap{ municipality =>
+
+      val municipalityUsers = users.filter( _.isAuthorizedToWrite(municipality) )
+                                    .filterNot( user => operatorUsersUsername.contains(user.username) )
+                                    .map( _.username )
+
+      val linearAssets = extractData(municipality)
+
+      if ( linearAssets != null && linearAssets.nonEmpty) {
+        val linearAssetsGroupedBy = linearAssets.groupBy(_.assetType)
+
+        assetTypesList.flatMap { assetType =>
+          if (linearAssetsGroupedBy.contains(assetType)) {
+            val currentAssetType = linearAssetsGroupedBy(assetType)
+
+            val (operatorUser, operatorDate) = extractUserAndDateTime(currentAssetType, operatorUsersUsername)
+            val (municipalityUser, municipalityDate) = extractUserAndDateTime(currentAssetType, municipalityUsers)
+
+            val firstElem = currentAssetType.head
+
+            Some( ExportAssetReport(municipalitiesInfo.filter(_.id == municipality).head.name, firstElem.assetType, currentAssetType.size,
+                              firstElem.assetNameFI, firstElem.assetGeometryType, operatorUser, operatorDate, municipalityUser, municipalityDate) )
+          }
+          else
+            None
+        }
+
+      }
+      else
+        None
```

Comment:
Maybe we can use some ideas of the suggestion but not as is.
Because if a municipality don't have an asset in the list of the extraction we don't want to send the information of that asset with value 0.

Reply:
As previously said, partial of suggestion implemented.
The reason for that is not to send rows with value zero because in the user story the example seems in that way to not send zeros

Reply:
Refactored the code